pydantic_ai_slim/pydantic_ai/toolsets/fastmcp.py

```diff
@@ -0,0 +1,231 @@
+        ImageContent,
+        TextContent,
+        Tool as MCPTool,
+    )
+
+    from pydantic_ai.mcp import TOOL_SCHEMA_VALIDATOR
+
+except ImportError as _import_error:
+    raise ImportError(
+        'Please install the `fastmcp` package to use the FastMCP server, '
+        'you can use the `fastmcp` optional group — `pip install "pydantic-ai-slim[fastmcp]"`'
+    ) from _import_error
+
+
+if TYPE_CHECKING:
+    from fastmcp.client.client import CallToolResult
+
+
+FastMCPToolResult = messages.BinaryContent | dict[str, Any] | str | None
+
+ToolErrorBehavior = Literal['model_retry', 'error']
+
+
+@dataclass
+class FastMCPToolset(AbstractToolset[AgentDepsT]):
+    """A FastMCP Toolset that uses the FastMCP Client to call tools from a local or remote MCP Server.
+
+    The Toolset can accept a FastMCP Client, a FastMCP Transport, or any other object which a FastMCP Transport can be created from.
+
+    See https://gofastmcp.com/clients/transports for a full list of transports available.
+    """
+
+    client: Client[Any]
+    """The FastMCP transport to use. This can be a local or remote MCP Server configuration, a transport string, or a FastMCP Client."""
+
+    _: KW_ONLY
+
+    tool_error_behavior: Literal['model_retry', 'error'] = field(default='error')
+    """The behavior to take when a tool error occurs."""
+
+    max_retries: int = field(default=2)
+    """The maximum number of retries to attempt if a tool call fails."""
+
+    _id: str | None = field(default=None)
+
+    @overload
+    def __init__(
+        self,
+        *,
+        client: Client[Any],
+        max_retries: int = 2,
+        tool_error_behavior: Literal['model_retry', 'error'] = 'error',
+        id: str | None = None,
+    ) -> None: ...
+
+    @overload
+    def __init__(
+        self,
+        transport: ClientTransport
+        | FastMCP
+        | FastMCP1Server
+        | AnyUrl
+        | Path
+        | MCPConfig
+        | dict[str, Any]
+        | str
+        | None = None,
+        *,
+        max_retries: int = 2,
+        tool_error_behavior: Literal['model_retry', 'error'] = 'error',
+        id: str | None = None,
+    ) -> None: ...
+
+    def __init__(
+        self,
+        transport: ClientTransport
+        | FastMCP
+        | FastMCP1Server
+        | AnyUrl
+        | Path
+        | MCPConfig
+        | dict[str, Any]
+        | str
+        | None = None,
+        *,
+        client: Client[Any] | None = None,
+        max_retries: int = 2,
+        tool_error_behavior: Literal['model_retry', 'error'] = 'error',
+        id: str | None = None,
+    ) -> None:
+        if not client and not transport:
+            raise ValueError('Either client or transport must be provided')
+
+        if client and transport:
+            raise ValueError('Either client or transport must be provided, not both')
+
+        if client:
+            self.client = client
+        else:
+            self.client = Client[Any](transport=transport)
```

Comment:
I'm guessing this will raise an error if the `dict[str,Any]` or `str` doesn't look like it should?

Reply:
yes, I can add tests for this, most common error raised will be a `ValueError(f"Could not infer a valid transport from: {transport}")`